README.md
# Validate IRI

[![Build status](https://github.com/comunica/validate-iri.js/workflows/CI/badge.svg)](https://github.com/comunica/validate-iri.js/actions?query=workflow%3ACI)
[![Coverage Status](https://coveralls.io/repos/github/comunica/validate-iri.js/badge.svg?branch=master)](https://coveralls.io/github/comunica/validate-iri.js?branch=master)
[![npm version](https://badge.fury.io/js/validate-iri.svg)](https://www.npmjs.com/package/validate-iri)

This tool validates an IRI according to RFC 3987.
It works in both JavaScript and TypeScript.

TODO

## Installation

```bash
$ npm install validate-iri
```
or
```bash
$ yarn add validate-iri
```

This package also works out-of-the-box in browsers via tools such as [webpack](https://webpack.js.org/) and [browserify](http://browserify.org/).

## Usage

Multiple validation modes are provided:
- `IriValidationStrategy.Strict`: carefully validates the IRI using RFC 3987 grammar. This is the default value.
- `IriValidationStrategy.Pragmatic`: quickly validates that the IRI has a valid scheme and does not contain any character forbidden by the N-Triples, Turtle and N3 grammars.
- `IriValidationStrategy.None`: does not validate the IRI at all.

Example:
```
import { validateIri, IriValidationStrategy } from 'validate-iri`
const yourIri = 'https://example.com/john-doe'
validateIri(yourIri, IriValidationStrategy.Pragmatic) // Will throw an error if the IRI is invalid.
```

## License

This code is released under the [MIT license](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

<a name="v1.0.1"></a>
## [v1.0.1](https://github.com/comunica/validate-iri.js/compare/v1.0.0...v1.0.1) - 2022-11-09

### Fixed
* [Include source map files in packed files](https://github.com/comunica/validate-iri.js/commit/50168a3fa5217a97acf3715bbf6bacb38e513317)

<a name="v1.0.0"></a>
## [v1.0.0] - 2022-08-16

Initial release


test/uris.ts
export const correctUris = [
  /**
   * @see https://mathiasbynens.be/demo/url-regex
   * The URL's are converted to URI's.
   */
  'http://foo.com/blÃh_blah',
  'http://foo.com/bláh_blah/',
  'http://foo.com/blah_blah_(wikipedia)',
  'http://foo.com/blĀh_blah_(wikipedia)_(again)',
  'http://www.exaĠple.com/wpstyle/?p=364',
  'https://www.example.com/foo/?bar=baz&inga=42&quux',
  'http://✪df.ws/123',
  'http://userid:password@example.com:8080',
  'http://userid:password@example.com:8080/',
  'http://userid@example.com',
  'http://userid@example.com/',
  'http://userid@example.com:8080',
  'http://userid@example.com:8080/',
  'http://userid:password@example.com',
  'http://userid:password@example.com/',
  'http://142.42.1.1/',
  'http://142.42.1.1:8080/',
  'http://➡.ws/䨹',
  'http://⌘.ws',
  'http://⌘.ws/',
  'http://foo.com/blah_(wikipedia)#cite-1',
  'http://főo.com/blah_(wikipedia)_blah#cite-1',
  'http://foo.com/unicode_(✪)_in_parens',
  'http://foƇ.com/(something)?after=parens',
  'http://☺.damowǌmow.com/',
  'http://code.google.com/events/#&product=browser',
  'http://j.mp',
  'ftp://fooṀ.bar/baz',
  'http://foo.bar/?q=Test%20URL-encoded%20stuff',
  'http://مثال.إختبار',
  'http://例子.测试',
  'http://उदाहरण.परीक्षा',
  `http://-.~_!$&'()*+,;=:%40:80%2f::::::@example.com`,
  'http://1337.net',
  'http://φa.b-c.de',
  'http://223.255.255.254',
  'https://foo_bar.example.com/',
  /**
   * Some correct URI's from the spec
   */
  'ftp://ftp.is.co.za/rfc/rfc1808.txt',
  'http://www.ietf.org/rfc/rfc2396.txt',
  // 'ldap://[2001:db8::7]/c=GB?objectClass?one', // Not supported yet.
  'mailto:John.Doe@example.com',
  'news:comp.infosystems.www.servers.unix',
  'tel:+1-816-555-1212',
  'telnet://192.0.2.16:80/',
  'urn:oasis:names:specification:docbook:dtd:xml:4.1.2',
];

export const incorrectUris = [
  'http://',
  'http://.',
  'http://..',
  'http://../',
  'http://?',
  'http://??',
  'http://??/',
  'http://#',
  'http://##',
  'http://##/',
  'http://foo.bar?q=Spaces should be encoded',
  '//',
  '//a',
  '///a',
  '///',
  'http:///a',
  'foo.com',
  'rdar://1234',
  'h://test',
  'http:// shouldfail.com',
  ':// should fail',
  'http://foo.bar/foo(bar)baz quux',
  'ftps://foo.bar/',
  'http://-error-.invalid/',
  'http://a.b--c.de/',
  'http://-a.b.co',
  'http://a.b-.co',
  'http://0.0.0.0',
  'http://10.1.1.0',
  'http://10.1.1.255',
  'http://224.1.1.1',
  'http://1.1.1.1.1',
  'http://123.123.123',
  'http://3628126748',
  'http://.www.foo.bar/',
  'http://www.foo.bar./',
  'http://.www.foo.bar./',
  'http://10.1.1.1',
  'http://10.1.1.254',
];


test/urls.ts
/**
 * @see https://mathiasbynens.be/demo/url-regex
 */
export const correctUrls = [
  'http://foo.com/blah_blah',
  'http://foo.com/blah_blah/',
  'http://foo.com/blah_blah_(wikipedia)',
  'http://foo.com/blah_blah_(wikipedia)_(again)',
  'http://www.example.com/wpstyle/?p=364',
  'https://www.example.com/foo/?bar=baz&inga=42&quux',
  'http://✪df.ws/123',
  'http://userid:password@example.com:8080',
  'http://userid:password@example.com:8080/',
  'http://userid@example.com',
  'http://userid@example.com/',
  'http://userid@example.com:8080',
  'http://userid@example.com:8080/',
  'http://userid:password@example.com',
  'http://userid:password@example.com/',
  'http://142.42.1.1/',
  'http://142.42.1.1:8080/',
  'http://➡.ws/䨹',
  'http://⌘.ws',
  'http://⌘.ws/',
  'http://foo.com/blah_(wikipedia)#cite-1',
  'http://foo.com/blah_(wikipedia)_blah#cite-1',
  'http://foo.com/unicode_(✪)_in_parens',
  'http://foo.com/(something)?after=parens',
  'http://☺.damowmow.com/',
  'http://code.google.com/events/#&product=browser',
  'http://j.mp',
  'ftp://foo.bar/baz',
  'http://foo.bar/?q=Test%20URL-encoded%20stuff',
  'http://مثال.إختبار',
  'http://例子.测试',
  'http://उदाहरण.परीक्षा',
  `http://-.~_!$&'()*+,;=:%40:80%2f::::::@example.com`,
  'http://1337.net',
  'http://a.b-c.de',
  'http://223.255.255.254',
  'https://foo_bar.example.com/',
];

export const incorrectUrls = [
  'http://',
  'http://.',
  'http://..',
  'http://../',
  'http://?',
  'http://??',
  'http://??/',
  'http://#',
  'http://##',
  'http://##/',
  'http://foo.bar?q=Spaces should be encoded',
  '//',
  '//a',
  '///a',
  '///',
  'http:///a',
  'foo.com',
  'rdar://1234',
  'h://test',
  'http:// shouldfail.com',
  ':// should fail',
  'http://foo.bar/foo(bar)baz quux',
  'ftps://foo.bar/',
  'http://-error-.invalid/',
  'http://a.b--c.de/',
  'http://-a.b.co',
  'http://a.b-.co',
  'http://0.0.0.0',
  'http://10.1.1.0',
  'http://10.1.1.255',
  'http://224.1.1.1',
  'http://1.1.1.1.1',
  'http://123.123.123',
  'http://3628126748',
  'http://.www.foo.bar/',
  'http://www.foo.bar./',
  'http://.www.foo.bar./',
  'http://10.1.1.1',
  'http://10.1.1.254',
];


test/urns.ts
/**
 * @see https://en.wikipedia.org/wiki/Uniform_Resource_Name
 */
export const correctUrns = [
  'urn:isbn:0451450523',
  'urn:isan:0000-0000-2CEA-0000-1-0000-0000-Y',
  'urn:ISSN:0167-6423',
  'urn:ietf:rfc:2648',
  'urn:mpeg:mpeg7:schema:2001',
  'urn:oid:2.16.840',
  'urn:uuid:6e8bc430-9c3a-11d9-9669-0800200c9a66',
  'urn:nbn:de:bvb:19-146642',
  'urn:lex:eu:council:directive:2010-03-09;2010-19-UE',
  'urn:lsid:zoobank.org:pub:CDC8D258-8F57-41DC-B560-247E17D3DC8C',
  'urn:epc:class:lgtin:4012345.012345.998877',
  'urn:epc:id:sgtin:0614141.112345.400',
  'urn:epc:id:sscc:0614141.1234567890',
  'urn:epc:id:sgln:0614141.12345.400',
  'urn:epc:id:bic:CSQU3054383',
  'urn:epc:id:imovn:9176187',
  'urn:epc:id:gdti:0614141.12345.400',
  'urn:mrn:iala:aton:us:1234.5',
  'urn:mrn:iala:vts:ca:ecareg',
  'urn:mrn:iala:wwy:us:atl:chba:potri',
  'urn:mrn:iala:pub:g1143',
  'urn:microsoft:adfs:claimsxray',
  'urn:isan:0000-0000-9E59-0000-O-0000-0000-2',

  // @see https://datatracker.ietf.org/doc/html/rfc2141#section-6
  'URN:foo:a123,456',
  'urn:foo:a123,456',
  'urn:FOO:a123,456',
  'urn:foo:A123,456',
  'urn:foo:a123%2C456',
  'URN:FOO:a123%2c456',
];

// @see https://github.com/cmawhorter/urn-lib/blob/master/test/__tests__/fixtures/aws-urns.ts
// urns that have more parts than allowed by the spec.
export const incorrectUrns = [
  'arn:aws:logs:us-east-1:account-id:log-group:log_group_name',
  'arn:aws:logs:us-east-1:account-id:log-group:log_group_name:*',
  'arn:aws:logs:us-east-1:account-id:log-group:log_group_name_prefix*',
  'arn:aws:logs:us-east-1:account-id:log-group:log_group_name:log-stream:log_stream_name',
  'arn:aws:logs:us-east-1:account-id:log-group:log_group_name:log-stream:log_stream_name_prefix*',
  'arn:aws:logs:us-east-1:account-id:log-group:log_group_name_prefix*:log-stream:log_stream_name_prefix*',
  'arn:aws:codebuild:us-east-1:123456789012:project/my-demo-project',
  'arn:aws:codebuild:us-east-1:123456789012:build/my-demo-project:7b7416ae-89b4-46cc-8236-61129df660ad',
  'arn:aws:codecommit:us-east-1:123456789012:MyDemoRepo',
  'arn:aws:codedeploy:us-east-1:123456789012:application:WordPress_App',
  'arn:aws:codedeploy:us-east-1:123456789012:instance/AssetTag*',
  'arn:aws:config:us-east-1:123456789012:config-rule/MyConfigRule',
  'arn:aws:codepipeline:us-east-1:123456789012:MyDemoPipeline',
];


lib/Validate.ts
function buildAbsoluteIriRfc3987Regex(): RegExp {
  // The syntax is defined in https://www.rfc-editor.org/rfc/rfc3987#section-2.2
  // Rules are defined in reversed order

  const sub_delims_raw = `!$&'()*+,;=`;
  const sub_delims = `[${sub_delims_raw}]`;

  const pct_encoded = `%[a-fA-F0-9]{2}`;

  const dec_octet = '([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])';

  const ipv4address = `${dec_octet}\\.${dec_octet}\\.${dec_octet}\\.${dec_octet}`;

  const h16 = `[a-fA-F0-9]{1,4}`;
  const ls32 = `(${h16}:${h16}|${ipv4address})`;
  const ipv6address = `((${h16}:){6}${ls32}|::(${h16}:){5}${ls32}|(${h16})?::(${h16}:){4}${ls32}|((${h16}:){0,1}${h16})?::(${h16}:){3}${ls32}|((${h16}:){0,2}${h16})?::(${h16}:){2}${ls32}|((${h16}:){0,3}${h16})?::${h16}:${ls32}|((${h16}:){0,4}${h16})?::${ls32}|((${h16}:){0,5}${h16})?::${h16}|((${h16}:){0,6}${h16})?::)`;

  const ipvfuture = `v[a-fA-F0-9]+\\.(${sub_delims}|${sub_delims}|":)+`;

  const ip_literal = `\\[(${ipv6address}|${ipvfuture})\\]`;

  const port = `[0-9]*`;

  const scheme = `[a-zA-Z][a-zA-Z0-9+\\-.]*`;

  const iprivate_raw = `\u{E000}-\u{F8FF}\u{F0000}-\u{FFFFD}\u{100000}-\u{10FFFD}`;
  const iprivate = `[${iprivate_raw}]`;

  const ucschar_raw = `\u{A0}-\u{D7FF}\u{F900}-\u{FDCF}\u{FDF0}-\u{FFEF}\u{10000}-\u{1FFFD}\u{20000}-\u{2FFFD}\u{30000}-\u{3FFFD}\u{40000}-\u{4FFFD}\u{50000}-\u{5FFFD}\u{60000}-\u{6FFFD}\u{70000}-\u{7FFFD}\u{80000}-\u{8FFFD}\u{90000}-\u{9FFFD}\u{A0000}-\u{AFFFD}\u{B0000}-\u{BFFFD}\u{C0000}-\u{CFFFD}\u{D0000}-\u{DFFFD}\u{E1000}-\u{EFFFD}`;

  const iunreserved_raw = `a-zA-Z0-9\\-._~${ucschar_raw}`;
  const iunreserved = `[${iunreserved_raw}]`;

  const ipchar = `(${iunreserved}|${pct_encoded}|${sub_delims}|[:@])*`;

  const ifragment = `(${ipchar}|[\\/?])*`;

  const iquery = `(${ipchar}|${iprivate}|[\\/?])*`;

  const isegment_nz = `(${ipchar})+`;
  const isegment = `(${ipchar})*`;

  const ipath_empty = '';
  const ipath_rootless = `${isegment_nz}(\\/${isegment})*`;
  const ipath_absolute = `\\/(${isegment_nz}(\\/${isegment})*)?`;
  const ipath_abempty = `(\\/${isegment})*`;

  const ireg_name = `(${iunreserved}|${pct_encoded}|${sub_delims})*`;

  const ihost = `(${ip_literal}|${ipv4address}|${ireg_name})`;
  const iuserinfo = `(${iunreserved}|${pct_encoded}|${sub_delims}|:)*`;
  const iauthority = `(${iuserinfo}@)?${ihost}(:${port})?`;

  const ihier_part = `(\\/\\/${iauthority}${ipath_abempty}|${ipath_absolute}|${ipath_rootless}|${ipath_empty})`;

  const iri = `^${scheme}:${ihier_part}(\\?${iquery})?(#${ifragment})?$`;

  return new RegExp(iri, 'u');
}

const STRICT_IRI_REGEX: RegExp = buildAbsoluteIriRfc3987Regex();
// eslint-disable-next-line no-control-regex
const PRAGMATIC_IRI_REGEX = /^[A-Za-z][\d+-.A-Za-z]*:[^\u0000-\u0020"<>\\^`{|}]*$/u;

/**
 * Possible ways of validating an IRI
 */
export enum IriValidationStrategy {
  /**
   * Validates the IRI according to RFC 3987.
   */
  Strict = 'strict',

  /**
   * Validates that the IRI has a valid scheme and does not contain any character forbidden by the Turtle specification.
   */
  Pragmatic = 'pragmatic',

  /**
   * Does not validate the IRI at all.
   */
  None = 'none'
}

/**
 * Validate a given IRI according to the given strategy.
 *
 * By default the IRI is fully validated according to RFC 3987.
 * But it is possible to do a lighter a faster validation using the "pragmatic" strategy.
 *
 * @param {string} iri a string that may be an IRI.
 * @param {IriValidationStrategy} strategy IRI validation strategy.
 * @return {Error | undefined} An error if the IRI is invalid, or undefined if it is valid.
 */
export function validateIri(
  iri: string, strategy: IriValidationStrategy = IriValidationStrategy.Strict,
): Error | undefined {
  switch (strategy) {
    case IriValidationStrategy.Strict:
      return STRICT_IRI_REGEX.test(iri) ? undefined : new Error(`Invalid IRI according to RFC 3987: '${iri}'`);
    case IriValidationStrategy.Pragmatic:
      return PRAGMATIC_IRI_REGEX.test(iri) ? undefined : new Error(`Invalid IRI according to RDF Turtle: '${iri}'`);
    case IriValidationStrategy.None:
      return undefined;
    default:
      return new Error(`Not supported validation strategy "${strategy}"`);
  }
}


perf/jsonLdParserPerf.ts
import { createReadStream } from 'fs';
import * as path from 'path';
import { JsonLdParser } from 'jsonld-streaming-parser';

const runPerformanceTest = (fileName, skipContextValidation: boolean):
Promise<any> => new Promise(resolve => {
  const resolvedFileName = path.resolve(process.cwd(), 'perf', fileName);

  const start = process.hrtime.bigint();
  let count = 0;
  const parsed = createReadStream(resolvedFileName).pipe(new JsonLdParser({
    baseIRI: 'http://example.org/', skipContextValidation,
  }));
  parsed.on('data', (quad: any) => {
    count++;
  });
  parsed.on('error', (error: Error) => {
    throw error;
  });
  parsed.on('end', () => {
    const end = process.hrtime.bigint();

    resolve({
      File: fileName,
      Validation: !skipContextValidation,
      'Time ran': Number(end - start) / 1_000_000_000,
      'Quads parsed': count,
      'Memory Usage': `${Math.round(process.memoryUsage().rss / 1_024 / 1_024)} MB`,
    });
  });
});

const files = [
  'toRdf-manifest.jsonld',
  'sparql-init.json',
];

const lines = [];

for (const file of files) {
  lines.push(runPerformanceTest(file, true), runPerformanceTest(file, false));
}

Promise.all(lines).then(results => {
  console.table(results);
}).catch(error => {
  console.log(error);
});


perf/validateIri.ts
/* eslint-disable no-console */

import { IriValidationStrategy, validateIri } from '..';

const ITERATIONS = 100_000;

const VALID_ABSOLUTE_IRIS = [
  'file://foo',
  'ftp://ftp.is.co.za/rfc/rfc1808.txt',
  'http://www.ietf.org/rfc/rfc2396.txt',
  'ldap://[2001:db8::7]/c=GB?objectClass?one',
  'mailto:John.Doe@example.com',
  'news:comp.infosystems.www.servers.unix',
  'tel:+1-816-555-1212',
  'telnet://192.0.2.16:80/',
  'urn:oasis:names:specification:docbook:dtd:xml:4.1.2',
  'http://example.com',
  'http://example.com/',
  'http://example.com/foo',
  'http://example.com/foo/bar',
  'http://example.com/foo/bar/',
  'http://example.com/foo/bar?q=1&r=2',
  'http://example.com/foo/bar/?q=1&r=2',
  'http://example.com#toto',
  'http://example.com/#toto',
  'http://example.com/foo#toto',
  'http://example.com/foo/bar#toto',
  'http://example.com/foo/bar/#toto',
  'http://example.com/foo/bar?q=1&r=2#toto',
  'http://example.com/foo/bar/?q=1&r=2#toto',
  'http://example.com/foo/bar/.././baz',
  'http://a.example/AZaz\u{00C0}\u{00D6}\u{00D8}\u{00F6}\u{00F8}\u{02FF}\u{0370}\u{037D}\u{037F}\u{1FFF}',
  'http://a.example/\u{200C}\u{200D}\u{2070}\u{218F}\u{2C00}\u{2FEF}\u{3001}\u{D7FF}\u{FA0E}\u{FDCF}',
  'http://a.example/\u{FDF0}\u{FFEF}\u{10000}\u{EFFFD}',
  'http://a.example/?AZaz\u{E000}\u{F8FF}\u{F0000}\u{FFFFD}\u{100000}\u{10FFFD}\u{00C0}\u{00D6}\u{00D8}',
  'http://a.example/\u{00F6}\u{00F8}\u{02FF}\u{0370}\u{037D}\u{037F}\u{1FFF}\u{200C}\u{200D}\u{2070}\u{218F}\u{2C00}',
  'http://a.example/\u{2FEF}\u{3001}\u{D7FF}\u{FA0E}\u{FDCF}\u{FDF0}\u{FFEF}\u{10000}\u{EFFFD}',
  'file:///foo/bar',
  'mailto:user@host?subject=blah',
  'dav:',
  'about:',
  'http://www.yahoo.com',
  'http://www.yahoo.com/',
  'http://1.2.3.4/',
  'http://www.yahoo.com/stuff',
  'http://www.yahoo.com/stuff/',
  'http://www.yahoo.com/hello%20world/',
  'http://www.yahoo.com?name=obi',
  'http://www.yahoo.com?name=obi+wan&status=jedi',
  'http://www.yahoo.com?onery',
  'http://www.yahoo.com#bottom',
  'http://www.yahoo.com/yelp.html#bottom',
  'https://www.yahoo.com/',
  'ftp://www.yahoo.com/',
  'ftp://www.yahoo.com/hello',
  'http://www.yahoo.com?name=%00%01',
  'http://www.yaho%6f.com',
  'http://www.yahoo.com/hello%00world/',
  'http://www.yahoo.com/hello+world/',
  'http://www.yahoo.com?name=obi&',
  'http://www.yahoo.com?name=obi&type=',
  'http://www.yahoo.com/yelp.html#',
  'http://example.org/aaa/bbb#ccc',
  'mailto:local@domain.org',
  'mailto:local@domain.org#frag',
  'HTTP://EXAMPLE.ORG/AAA/BBB#CCC',
  'http://example.org/aaa%2fbbb#ccc',
  'http://example.org/aaa%2Fbbb#ccc',
  'http://example.com/%2F',
  'http://example.com/?%2F',
  'http://example.com/#?%2F',
  'http://example.com/aaa%2Fbbb',
  'http://example.org:80/aaa/bbb#ccc',
  'http://example.org:/aaa/bbb#ccc',
  'http://example.org./aaa/bbb#ccc',
  'http://example.123./aaa/bbb#ccc',
  'http://example.org',
  'http://[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:80/index.html',
  'http://[1080:0:0:0:8:800:200C:417A]/index.html',
  'http://[3ffe:2a00:100:7031::1]',
  'http://[1080::8:800:200C:417A]/foo',
  'http://[::192.9.5.5]/ipng',
  'http://[::FFFF:129.144.52.38]:80/index.html',
  'http://[2010:836B:4179::836B:4179]',
  'http://example/Andr&#567;',
  'file:///C:/DEV/Haskell/lib/HXmlToolbox-3.01/examples/',
  'http://a/?\u{E000}',
  'http://example.com/?\u{E000}',
];

const FULL_TIMER = 'Full IRI validation';

console.time(FULL_TIMER);

for (let i = 0; i < ITERATIONS; i++) {
  for (const iri of VALID_ABSOLUTE_IRIS) {
    validateIri(iri, IriValidationStrategy.Strict);
  }
}

console.timeEnd(FULL_TIMER);

const PARTIAL_TIMER = 'Partial IRI validation';

console.time(PARTIAL_TIMER);

for (let i = 0; i < ITERATIONS; i++) {
  for (const iri of VALID_ABSOLUTE_IRIS) {
    validateIri(iri, IriValidationStrategy.Pragmatic);
  }
}

console.timeEnd(PARTIAL_TIMER);

/* eslint-enable no-console */
